### DISCUSSION TITLE
I expected the type to be inferred as `SetAtom<...>`, but it is inferred as `never`.

### DISCUSSION THREAD

**Discussion by mjsdo** [pmndrs/jotai#1831] [Q&A]
I have an object with multiple types of atoms, and I created a custom hook using `useAtom(object[key])` to return `[state, setState]`.
However, the type of `state` is inferred correctly, but the type of `setState` is inferred as `never`.

Here is the code where the problem occurs.

```tsx
import { atom, useAtom } from "jotai";
import { ChangeEventHandler, useState } from "react";

const fieldAtoms = {
  username: atom(""),
  password: atom(""),
  checked: atom(false)
};

const useField = <T extends keyof typeof fieldAtoms>(prop: T) => {
  return useAtom(fieldAtoms[prop]);
};

const App = () => {
  // the type of username is correctly inferred as string,
  // but the type of setUsername is inferred as never.
  const [username, setUsername] = useField("username");

  const onChangeUsername: ChangeEventHandler<HTMLInputElement> = (e) => {
    setUsername(e.target.value);
  };

  return (
    // omit
  );
}
```

I wrote a sandbox for comparison with other cases. 
https://codesandbox.io/s/jotai-example-6i8neb?file=/src/App.tsx

To sum up the code in the sandbox,

1. If the object has same type of atoms, the problem does not occur. 
2. If the object has multiple types of atoms, the problem occurs.
3. The type is inferred as expected if I use literal rather than atom.

**Comment by dai-shi** [comment]
Oh, interesting. I thought it would work. Not sure if it's possible, but someone can feel free to dig into it.
